whole-tale/gwvolman
# Update DataONE Tale

## Background
There are a couple of changes....

1. While I was working around in the file, there were a few spots where I cleared the documentation up
2. I moved the resource map to the DataONEMetadata class
3. I moved the client object to the DataONEPublishProvider class
4. Removed some commented out code (we can't rollback failed submission)
5. Played with the tests
6. Updated the README

The changes are largely split into three commits (apologies): `Update Tales`, `Add Tests` and `Fix Tests`.

`Update Tales` has the logic for updating Tales that were already published and handles appending DataCite:IsDerivedFrom if the Tale was copied from a published Tale. A few updates to it were made in `Add Tests`. `Add Tests` was the first pass at expanding the unit tests. Inside you'll see that the TALE and MANIFEST variables are in each file. The `Fix Tests` commit largely moves these to `__init__.py`.



### Misc

There are a few miscellaneous changes that range from marking some methods as `static`, rewording some of the exceptions, improving the docstrings and type annotating types from the dataone library for easier reading & IDE support.

The styling changes in 8c3bbf9 are fairly trivial. The riskiest one is probably renaming the `zip` variable (shadows the `zip` package) to `zip_file`.

### Updating

A Tale is 'updated' when it's published a second time. If a Tale is published to DataONE, and then a copy is made and published- `datacite:IsDerivedFrom` pointing to the first Tale's doi is added to the resource map. The relevant code for this is [here](https://github.com/whole-tale/gwvolman/pull/111/files#diff-fb3f74bc232977b68ea420c16a6b6d7aR118)

When a Tale is published to DataONE two times, the previous system metadata is retrieved. A new system metadata document is made, but replaces values such as md5 and size from the copied one. This system metadata document is for the object that will be obsoleting the one in DataONE.

`[_obsolete_object]()` just calls the DataONE python library's update method which takes care of the rest of the heavy lifting. 


## Tests
I restructured the unit tests to match the src directory. I took the json from the `test_publish.py` and moved it into the `__init__.py` file so that they can be used in other tests. You can also see that I have added structures for published tales, etc.  Then I pulled the dataone logic of `test_publish.py` and moved it over into `test_dataone_publish.py`. I added `test_dataone_metadata.py` to test some of the methods in the `metadata.py` file but left out tests for the EML.


## Testing

1. Create a Tale
2. Publish it to DataONE
3. View the landing page for that Tale (don't close the tab!)
4. Publish again
5. Refresh the URL from step 3, you should see a notification that there is an updated version available


_

1. Create a Tale
2. Publish it to DataONE
3. Make the Tale public
4. Sign out and sign in under a new account
5. Copy the Tale from step 1
6. Publish it to DataONE
7. Copy the ID of the resource map (the uuid in the picture below)
8. Visit https://dev.nceas.ucsb.edu/knb/d1/mn/v2/object/<RESOURCE_MAP_ID>
9. Search for "datacite"
10. Check that it references the copied Tale's doi


Example:

I published a Tale [here](https://dev.nceas.ucsb.edu/view/urn:uuid:12539245-06fc-482e-bec4-da2db24490c2), copied it and republished [here](https://dev.nceas.ucsb.edu/view/urn:uuid:e3cb4097-4182-4615-a8c4-74c91cb8a928). In the _new_ resource map found [here]( https://dev.nceas.ucsb.edu/knb/d1/mn/v2/object/urn:uuid:e3cb4097-4182-4615-a8c4-74c91cb8a928), you can see that the datacite:IsDerivedFrom points to the DOI of the first package.




Find the resource map pid at the top of the file table on the dataset landing page, shown below.
<img width="1640" alt="Screen Shot 2020-09-16 at 5 03 20 PM" src="https://user-images.githubusercontent.com/9289652/93404486-2b38eb00-f83f-11ea-8365-a0b41e96be85.png">
